lms/djangoapps/grades/management/commands/tests/test_reset_grades.py

```diff
@@ -0,0 +1,289 @@
+                    PersistentSubsectionGrade.update_or_create_grade(**subsection_grade_params)
+
+    def _assert_grades_exist_for_courses(self, course_keys):
+        """
+        Assert grades for given courses exist.
+        """
+        for course_key in course_keys:
+            self.assertIsNotNone(PersistentCourseGrade.read_course_grade(self.user_ids[0], course_key))
+            for subsection_key in self.subsection_keys_by_course[course_key]:
+                self.assertIsNotNone(PersistentSubsectionGrade.read_grade(self.user_ids[0], subsection_key))
+
+    def _assert_grades_absent_for_courses(self, course_keys):
+        """
+        Assert grades for given courses do not exist.
+        """
+        for course_key in course_keys:
+            with self.assertRaises(PersistentCourseGrade.DoesNotExist):
+                PersistentCourseGrade.read_course_grade(self.user_ids[0], course_key)
+
+            for subsection_key in self.subsection_keys_by_course[course_key]:
+                with self.assertRaises(PersistentSubsectionGrade.DoesNotExist):
+                    PersistentSubsectionGrade.read_grade(self.user_ids[0], subsection_key)
+
+    def _assert_stat_logged(self, mock_log, num_rows, grade_model_class, message_substring, log_offset):
+        self.assertIn('reset_grade: ' + message_substring, mock_log.info.call_args_list[log_offset][0][0])
+        self.assertEqual(grade_model_class.__name__, mock_log.info.call_args_list[log_offset][0][1])
+        self.assertEqual(num_rows, mock_log.info.call_args_list[log_offset][0][2])
+
+    def _assert_course_delete_stat_logged(self, mock_log, num_rows):
+        self._assert_stat_logged(mock_log, num_rows, PersistentCourseGrade, 'Deleted', log_offset=4)
+
+    def _assert_subsection_delete_stat_logged(self, mock_log, num_rows):
+        self._assert_stat_logged(mock_log, num_rows, PersistentSubsectionGrade, 'Deleted', log_offset=2)
+
+    def _assert_course_query_stat_logged(self, mock_log, num_rows, num_courses=None):
+        if num_courses is None:
+            num_courses = self.num_courses
+        log_offset = num_courses + 1 + num_courses + 1
+        self._assert_stat_logged(mock_log, num_rows, PersistentCourseGrade, 'Would delete', log_offset)
+
+    def _assert_subsection_query_stat_logged(self, mock_log, num_rows, num_courses=None):
+        if num_courses is None:
+            num_courses = self.num_courses
+        log_offset = num_courses + 1
+        self._assert_stat_logged(mock_log, num_rows, PersistentSubsectionGrade, 'Would delete', log_offset)
+
+    def _date_from_now(self, days=None):
+        return datetime.now() + timedelta(days=days)
+
+    def _date_str_from_now(self, days=None):
+        future_date = self._date_from_now(days=days)
+        return future_date.strftime(reset_grades.DATE_FORMAT)
+
+    @patch('lms.djangoapps.grades.management.commands.reset_grades.log')
+    def test_reset_all_courses(self, mock_log):
```

Comment:
I was worried these were going to table scan to count, but innodb is counting an index (not the primary funnily)